SonarQube.Client/Requests/IRequestFactory.cs

```diff
@@ -0,0 +1,30 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarQube.Client.Requests
+{
+    /// <summary>
+    /// Factory that creates request instance appropriate to the supplied server type/version
+    /// </summary>
+    internal interface IRequestFactory
+    {
+        TRequest Create<TRequest>(ServerInfo serverInfo) where TRequest : IRequest;
+    }
```

Comment:
New abstraction to decouple the `ISonarQubeService`  from the single existing factory. The other important change is that the supplied parameter is now `ServerInfo` rather than just `Version`.

Logically, `ServerInfo` should contain all of the info a factory should need to decide to which API to call. In practice this means the version and whether the server is SQ or SC.